File: reference/ssklib/resolve.py
import copy

from .error import SSKError

_INHERITABLE = frozenset({
    'points', 'rotation', 'size', 'shape', 'sides',
    'mode', 'affects', 'properties',
})


def resolve(doc: dict, *, in_place: bool = True) -> dict:
    if not isinstance(doc, dict):
        raise SSKError("document must be a mapping")

    target = doc if in_place else copy.deepcopy(doc)
    pieces = target.get('pieces')
    if not isinstance(pieces, list):
        raise SSKError("document must contain a pieces list")

    ids = []
    for index, piece in enumerate(pieces):
        if not isinstance(piece, dict):
            raise SSKError(f"piece {index}: expected mapping")
        if 'id' not in piece:
            raise SSKError(f"piece {index}: missing id")
        pid = piece['id']
        if isinstance(pid, bool) or not isinstance(pid, int):
            raise SSKError(f"piece {index}: id must be an integer")
        ids.append(pid)

    if len(set(ids)) != len(ids):
        raise SSKError("piece ids must be unique before inheritance resolution")

    by_id = {p['id']: p for p in pieces}

    _check_inheritance_graph(by_id)

    done: set[int] = set()
    for pid in sorted(by_id):
        _do(pid, by_id, done)

    return target


def _check_inheritance_graph(by_id: dict):
    visiting: set[int] = set()
    visited: set[int] = set()

    def visit(pid: int):
        if pid in visited:
            return
        if pid in visiting:
            raise SSKError(f"circular inheritance involving piece {pid}")
        visiting.add(pid)
        piece = by_id[pid]
        if 'from' in piece:
            fid = piece['from']
            if isinstance(fid, bool) or not isinstance(fid, int):
                raise SSKError(f"piece {pid}: from must be an integer")
            if fid == pid:
                raise SSKError(f"piece {pid}: self-reference is invalid")
            if fid not in by_id:
                raise SSKError(f"piece {pid}: from references non-existent piece {fid}")
            if fid > pid:
                raise SSKError(f"piece {pid}: from must reference a lower id")
            visit(fid)
        visiting.remove(pid)
        visited.add(pid)

    for pid in sorted(by_id):
        visit(pid)


def _do(pid: int, by_id: dict, done: set):
    if pid in done:
        return
    piece = by_id[pid]
    if 'from' not in piece:
        done.add(pid)
        return
    fid = piece['from']
    if fid not in done:
        _do(fid, by_id, done)
    src = by_id[fid]
    for field in _INHERITABLE:
        if field not in piece and field in src:
            piece[field] = copy.deepcopy(src[field])
    done.add(pid)

File: implementation/packages/typescript/src/resolve.ts
import { SSKError } from './error.js';
import type { Piece, ResolvedDocument, SSKDocument } from './types.js';

const INHERITABLE = ['points', 'rotation', 'size', 'shape', 'sides', 'mode', 'affects', 'properties'] as const;

export function resolve(doc: SSKDocument, options: { inPlace?: boolean } = {}): ResolvedDocument {
  if (!doc || typeof doc !== 'object' || Array.isArray(doc)) throw new SSKError('document must be a mapping');
  const target = options.inPlace === false ? deepCopy(doc) : doc;
  if (!Array.isArray(target.pieces)) throw new SSKError('document must contain a pieces list');

  const ids: number[] = [];
  target.pieces.forEach((piece, index) => {
    if (!piece || typeof piece !== 'object' || Array.isArray(piece)) throw new SSKError(`piece ${index}: expected mapping`);
    if (!('id' in piece)) throw new SSKError(`piece ${index}: missing id`);
    if (typeof piece.id === 'boolean' || !Number.isInteger(piece.id)) throw new SSKError(`piece ${index}: id must be an integer`);
    ids.push(piece.id);
  });
  if (new Set(ids).size !== ids.length) throw new SSKError('piece ids must be unique before inheritance resolution');

  const byId = new Map<number, Piece>();
  for (const piece of target.pieces) byId.set(piece.id, piece);
  checkInheritanceGraph(byId);

  const done = new Set<number>();
  for (const id of [...byId.keys()].sort((a, b) => a - b)) doResolve(id, byId, done);
  return target as ResolvedDocument;
}

function checkInheritanceGraph(byId: Map<number, Piece>): void {
  const visiting = new Set<number>();
  const visited = new Set<number>();

  function visit(pid: number): void {
    if (visited.has(pid)) return;
    if (visiting.has(pid)) throw new SSKError(`circular inheritance involving piece ${pid}`);
    visiting.add(pid);
    const piece = byId.get(pid)!;
    if ('from' in piece) {
      const fid = piece.from;
      if (typeof fid === 'boolean' || !Number.isInteger(fid)) throw new SSKError(`piece ${pid}: from must be an integer`);
      if (fid === pid) throw new SSKError(`piece ${pid}: self-reference is invalid`);
      if (!byId.has(fid!)) throw new SSKError(`piece ${pid}: from references non-existent piece ${fid}`);
      if (fid! > pid) throw new SSKError(`piece ${pid}: from must reference a lower id`);
      visit(fid!);
    }
    visiting.delete(pid);
    visited.add(pid);
  }

  for (const id of [...byId.keys()].sort((a, b) => a - b)) visit(id);
}

function doResolve(pid: number, byId: Map<number, Piece>, done: Set<number>): void {
  if (done.has(pid)) return;
  const piece = byId.get(pid)!;
  if (!('from' in piece)) {
    done.add(pid);
    return;
  }
  const fid = piece.from!;
  if (!done.has(fid)) doResolve(fid, byId, done);
  const source = byId.get(fid)!;
  for (const field of INHERITABLE) {
    if (!(field in piece) && field in source) {
      (piece as Record<string, unknown>)[field] = deepCopy((source as Record<string, unknown>)[field]);
    }
  }
  done.add(pid);
}

export function deepCopy<T>(value: T): T {
  return typeof structuredClone === 'function' ? structuredClone(value) : JSON.parse(JSON.stringify(value));
}
